### PR TITLE
generate types.json

### LINKED ISSUE #2062: Publish types.json with releases
Subscan and others require a types.json be published. We can generate one with the API-Augment code, but we don't publish it anywhere.

## AC
- It should be published as part of the api-augment
- It should be published as part of the release

## References
- Example from Crust: https://github.com/crustio/crust.js/blob/mainnet/packages/type-definitions/src/json/types.json

### PR DESCRIPTION
**Pull request by shannonwells**
# Goal
The goal of this PR is to generate a types.json file to help with being on Subscan et al.

Closes #2062 

# Checklist
- [x] Updated js/api-augment for Custom RPC OR Runtime API?
- [x] Added types.json generation to release.yml
- [x] Added types.json generation to package.json (two scripts, but postbuild doesn't seem to be used anywhere).
- [x] Added types.json generation to generate_js_definitions.js script so it is triggered by `make js`

### Verify
1. Check out this branch and run `make js`, which replicates the two needed steps run in CI.
2. if you `cd js/api-augment/dist`, this should exist: `json/types.json`
3. The content of `types.json` should look like the following, so that it would be included in the tarball if you run `npm pack`:
```json
{
    "RpcEvent": {
        "phase": "Option<u32>",
        "pallet": "u8",
        "event": "u8",
        "data": "Vec<u8>"
    },
    "HandleSuffix": "u16",
    "HandleResponse": {
        "base_handle": "String",
        "canonical_base": "String",
        "suffix": "u16"
    },
    "PresumptiveSuffixesResponse": {
        "suffixes": "Vec<HandleSuffix>",
        "base_handle": "String"
    },
    "BlockPaginationRequest": {
        "from_block": "BlockNumber",
        "from_index": "u32",
        "to_block": "BlockNumber",
        "page_size": "u32"
    },
    "MessageResponse": {
        "payload": "Option<Vec<u8>>",
        "cid": "Option<Vec<u8>>",
        "provider_msa_id": "MessageSourceId",
        "msa_id": "Option<MessageSourceId>",
        "index": "u16",
        "block_number": "BlockNumber",
        "payload_length": "Option<u32>"
    },
    "BlockPaginationResponseMessage": {
        "content": "Vec<MessageResponse>",
        "has_next": "bool",
        "next_block": "Option<BlockNumber>",
        "next_index": "Option<u32>"
    },
    "MessageSourceId": "u64",
    "DelegatorId": "MessageSourceId",
    "ProviderId": "MessageSourceId",
    "KeyInfoResponse": {
        "msa_keys": "Vec<AccountId>",
        "msa_id": "MessageSourceId"
    },
    "SchemaGrantResponse": {
        "schema_id": "SchemaId",
        "revoked_at": "BlockNumber"
    },
    "DelegationResponse": {
        "provider_id": "ProviderId",
        "permissions": "Vec<SchemaGrantResponse>"
    },
    "SchemaId": "u16",
    "SchemaModel": "Vec<u8>",
    "SchemaVersion": "u8",
    "SchemaResponse": {
        "schema_id": "SchemaId",
        "model": "SchemaModel",
        "model_type": "ModelType",
        "payload_location": "PayloadLocation",
        "settings": "Vec<SchemaSetting>"
    },
    "ModelType": {
        "_enum": [
            "AvroBinary",
            "Parquet"
        ]
    },
    "PayloadLocation": {
        "_enum": [
            "OnChain",
            "IPFS",
            "Itemized",
            "Paginated"
        ]
    },
    "SchemaSetting": {
        "_enum": [
            "AppendOnly",
            "SignatureRequired"
        ]
    },
    "SchemaVersionResponse": {
        "schema_name": "String",
        "schema_version": "SchemaVersion",
        "schema_id": "SchemaId"
    },
    "PageId": "u16",
    "PageHash": "u32",
    "PageNonce": "u16",
    "ItemizedStorageResponse": {
        "index": "u16",
        "payload": "Vec<u8>"
    },
    "PaginatedStorageResponse": {
        "page_id": "PageId",
        "msa_id": "MessageSourceId",
        "schema_id": "SchemaId",
        "content_hash": "PageHash",
        "nonce": "PageNonce",
        "payload": "Vec<u8>"
    },
    "ItemizedStoragePageResponse": {
        "msa_id": "MessageSourceId",
        "schema_id": "SchemaId",
        "content_hash": "PageHash",
        "nonce": "PageNonce",
        "items": "Vec<ItemizedStorageResponse>"
    }
}
```

### CODE REVIEW

**Review by enddynayn — APPROVED** [review]
👍 LGTM!

File: scripts/generate_js_definitions.sh [hunk 1]
```diff
@@ -41,6 +41,7 @@ else
     npm install # in case things have changed
     npm run fetch:local
     npm run build
+    npm run build:types-json
```
Comment (JoeCap08055) [inline scripts/generate_js_definitions.sh:44]: This seems redundant since `build:types-json` is now part of the `postbuild` script

**Review by wilwade — COMMENTED** [review]
Two notes about where the output publishes

File: scripts/generate_js_definitions.sh (hunk 1 above)
Comment (shannonwells) [reply scripts/generate_js_definitions.sh:44]: I'm not seeing where `postbuild` gets called, do you know what that script is for?

File: js/api-augment/package.json [hunk 2]
```diff
@@ -21,7 +21,8 @@
     "prepackage": "cp CONTRIBUTING.md README.md ../../LICENSE ./dist",
     "prebuild": "npm run generate",
     "build": "tsc -p ./tsconfig.types.json && tsc -p ./tsconfig.cjs.json && tsc -p ./tsconfig.esm.json",
-    "postbuild": "npm run package",
+    "build:types-json": "node scripts/generate.mjs",
+    "postbuild": "npm run package && npm run build:types-json",
```
[the remark this answers is not in the document]
Comment (shannonwells) [reply js/api-augment/package.json:25]: The way I understood it is that the scripts in the release steps are called from within the dist directory so they wouldn't need to be copied.

[the remark this answers is not in the document]
Comment (shannonwells) [reply js/api-augment/package.json:25]: `generate.mjs` output goes to [dist/json/types.json](https://github.com/frequency-chain/frequency/pull/2127/files#diff-c67d6a5af6908223c7eb2d1c7ea5df62a24622c2defc5fc9bd89ffae47eed374R11)

File: scripts/generate_js_definitions.sh (hunk 1 above)
Comment (JoeCap08055) [reply scripts/generate_js_definitions.sh:44]: The way npm scripts work, any script that starts with "pre" or "post" runs before/after the main script of the same name. So "postbuild" automatically runs after "build".

**Pull request closed without merging**